src/main.cc

```diff
@@ -1443,12 +1434,55 @@ ConfigureCurrentKid(const CommandLine &cmdLine)
     }
 }
 
-static void StartUsingConfig()
+static void
+ConfigureDebugging()
+{
+    if (opt_no_daemon) {
+        /* we have to init fdstat here. */
+        fd_open(0, FD_LOG, "stdin");
+        fd_open(1, FD_LOG, "stdout");
+        fd_open(2, FD_LOG, "stderr");
+    }
+    // we should not create cache.log outside chroot environment, if any
+    if (!Config.chroot_dir || Chrooted)
+        _db_init(Debug::cache_log, Debug::debugOptions);
```

Comment:
Assuming that Squid, sooner or later, must have a cache log, the `if` comment and `if` condition appear to imply that this function is called multiple times, and we wait for the right conditions (like Chrooted becoming true) to create cache.log. Is that what actually happens? Do we call this function more than once under some conditions?

Reply:
> I assume we may call this twice: firstly from the master process and secondly from the child.

Can you detail that a little please (in this review thread). For example:

When chrooting is configured:

1. Master process gets here once but does not call db_init() because Chrooted is false.
2. Child process gets here once and calls db_init() because Chrooted becomes true.

When chrooting is not configured:

1. Master process gets here once and calls db_init().
2. Child process gets here once and calls db_init() which repeats debugging initialization done in the earlier master process call above. However, that repetition is necessary because the two processes do not share variables (execvp() does not inherit parent process memory) and because the master process has to log something too.

The above texts are just _examples_ of the information I am looking for. I am not claiming that they reflect the actual Squid logic!

Reply:
Your description (for both chroot/no_chroot) is correct, i.e. reflects the actual Squid behavior.

Reply:
> Master process gets here once but does not call db_init()

This implies that the master process never initializes debugging when chrooting is configured. I assume that this problem is limited to SMP environments. I further assume the same problem exists in the official code. If all of the above is actually true, then please add an XXX to the official PR code. For example,

```
// XXX: With Config.chroot_dir set, SMP master process never calls db_init().
```

The XXX can go above the `if` statement in question.

The commit message should say that this bug exists in the official code.

What are the side effects of this bug? In SMP environment with chrooting configured/enabled, the master process never logs anything to the configured cache.log, right?

Reply:
>  In SMP environment with chrooting configured/enabled, the master process never logs anything to the configured cache.log, right?

Right.  I think that is the reason why in SMP mode watch_child() uses syslog() calls, rather than debugs().

Reply:
> add an XXX to the official PR code

Done at 164b96 (official PR474).